doc/man/gnunet-bcd.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd December 5, 2013
.Dt GNUNET-BCD 1
.Os
.Sh NAME
.Nm gnunet-bcd
.Nd run HTTP server to create GNS business cards
.Sh SYNOPSIS
.Nm
.Op Fl p Ar PORT | Fl -port= Ns Ar PORT
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
can be used to create an business card with a QR code containing the public key of a zone from the GNU Name System.
gnunet-bcd requires LaTeX (
.Xr pdflatex 1
) with various packages to be installed.
If it does not work for you, try installing the full TeXLive distribution first, for example if you are using the package-manager
.Xr apt 1
execute "apt-get install texlive-full".
.Bl -tag -width indent
.It Fl p Ar PORT | Fl -port= Ns Ar PORT
Run the HTTP server on port PORT.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl v | -version
Print GNUnet version number.
.El
.Sh SEE ALSO
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed Oct 30 10:50:00 UTC 2019
  Released gnunet 0.11.8
  -ng0

Sun Oct 27 17:17:00 2019 +0200
  Released gnunet 0.11.7
  -schanzen, ng0

Fri Oct 25 00:00:00 UTC 2019
  Dropped glpk dependency(-check). -ng0

Wed Oct 23 20:55:12 2019 +0200
  Make REST service start per user in a multiuser setup.
  -schanzen

Wed Oct 23 00:00:00 UTC 2019
  Remove setuid helpers. They never worked as intended. The
  fixed version is not portable and the defacto good way to
  handle this which is most portable is to let package managers
  handle this (as they do right now, the good ones), and let
  people handle this who read the documentation.
  This commit removes what would be patched out by the majority
  of package managers, which is setuid handling in Makefiles.
  It is very likely that no one will notice this code is gone.
  -ng0

Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression

src/gns/gnunet-bcd.c
/*
     This file is part of GNUnet.
     Copyright (C) 2013 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file gns/gnunet-bcd.c
 * @author Christian Grothoff
 * @brief HTTP server to create GNS business cards
 */

#include "platform.h"
#include <microhttpd.h>
#include "gnunet_util_lib.h"

/**
 * Error page to display if submitted GNS key is invalid.
 */
#define INVALID_GNSKEY \
  "<html><head><title>Error</title><body>Invalid GNS public key given.</body></html>"

/**
 * Error page to display on 404.
 */
#define NOT_FOUND \
  "<html><head><title>Error</title><body>404 not found</body></html>"

/**
 * Handle to the HTTP server as provided by libmicrohttpd
 */
static struct MHD_Daemon *daemon_handle;

/**
 * Our configuration.
 */
static const struct GNUNET_CONFIGURATION_Handle *cfg;

/**
 * Our primary task for the HTTPD.
 */
static struct GNUNET_SCHEDULER_Task *http_task;

/**
 * Our main website.
 */
static struct MHD_Response *main_response;

/**
 * Error: invalid gns key.
 */
static struct MHD_Response *invalid_gnskey_response;

/**
 * Error: 404
 */
static struct MHD_Response *not_found_response;

/**
 * Absolute name of the 'gns-bcd.tex' file.
 */
static char *resfile;

/**
 * Port number.
 */
static uint16_t port = 8888;


struct Entry
{
  const char *formname;
  const char *texname;
};


/**
 * Main request handler.
 */
static int
access_handler_callback (void *cls,
                         struct MHD_Connection *connection,
                         const char *url,
                         const char *method,
                         const char *version,
                         const char *upload_data,
                         size_t *upload_data_size,
                         void **con_cls)
{
  static int dummy;
  static const struct Entry map[] = { { "prefix", "prefix" },
                                      { "name", "name" },
                                      { "suffix", "suffix" },
                                      { "street", "street" },
                                      { "city", "city" },
                                      { "phone", "phone" },
                                      { "fax", "fax" },
                                      { "email", "email" },
                                      { "homepage", "homepage" },
                                      { "orga", "orga" },
                                      { "departmenti18n", "departmentde" },
                                      { "departmenten", "departmenten" },
                                      { "subdepartmenti18n",
                                        "subdepartmentde" },
                                      { "subdepartmenten", "subdepartmenten" },
                                      { "jobtitlei18n", "jobtitlegerman" },
                                      { "jobtitleen", "jobtitleenglish" },
                                      { "subdepartmenten", "subdepartmenten" },
                                      { NULL, NULL } };

  (void) cls;
  (void) version;
  (void) upload_data;
  (void) upload_data_size;
  if (0 != strcmp (method, MHD_HTTP_METHOD_GET))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_WARNING,
                _ ("Refusing `%s' request to HTTP server\n"),
                method);
    return MHD_NO;
  }
  if (NULL == *con_cls)
  {
    (*con_cls) = &dummy;
    GNUNET_log (GNUNET_ERROR_TYPE_DEBUG, "Sending 100 CONTINUE reply\n");
    return MHD_YES;   /* send 100 continue */
  }
  if (0 == strcasecmp (url, "/"))
    return MHD_queue_response (connection, MHD_HTTP_OK, main_response);
  if (0 == strcasecmp (url, "/submit.pdf"))
  {
    unsigned int i;
    char *p;
    char *tmp;
    char *deffile;
    struct GNUNET_CRYPTO_EcdsaPublicKey pub;
    size_t slen;
    FILE *f;
    struct stat st;
    struct MHD_Response *response;
    int fd;
    int ret;

    const char *gpg_fp = MHD_lookup_connection_value (connection,
                                                      MHD_GET_ARGUMENT_KIND,
                                                      "gpgfingerprint");
    const char *gns_nick = MHD_lookup_connection_value (connection,
                                                        MHD_GET_ARGUMENT_KIND,
                                                        "gnsnick");
    const char *gnskey =
      MHD_lookup_connection_value (connection, MHD_GET_ARGUMENT_KIND, "gnskey");
    if ((NULL == gnskey) ||
        (GNUNET_OK !=
         GNUNET_CRYPTO_ecdsa_public_key_from_string (gnskey,
                                                     strlen (gnskey),
                                                     &pub)))
    {
      return MHD_queue_response (connection,
                                 MHD_HTTP_OK,
                                 invalid_gnskey_response);
    }
    tmp = GNUNET_DISK_mkdtemp (gnskey);
    if (NULL == tmp)
    {
      GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "mktemp", gnskey);
      return MHD_NO;
    }
    GNUNET_asprintf (&deffile, "%s%s%s", tmp, DIR_SEPARATOR_STR, "def.tex");
    f = fopen (deffile, "w");
    if (NULL == f)
    {
      GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "open", deffile);
      GNUNET_free (deffile);
      GNUNET_DISK_directory_remove (tmp);
      GNUNET_free (tmp);
      return MHD_NO;
    }
    for (i = 0; NULL != map[i].formname; i++)
    {
      const char *val = MHD_lookup_connection_value (connection,
                                                     MHD_GET_ARGUMENT_KIND,
                                                     map[i].formname);
      if (NULL != val)
        fprintf (f, "\\def\\%s{%s}\n", map[i].texname, val);
      else
        fprintf (f, "\\def\\%s{}\n", map[i].texname);
    }
    if (NULL != gpg_fp)
    {
      char *gpg1;
      char *gpg2;

      slen = strlen (gpg_fp);
      gpg1 = GNUNET_strndup (gpg_fp, slen / 2);
      gpg2 = GNUNET_strdup (&gpg_fp[slen / 2]);
      fprintf (f, "\\def\\gpglineone{%s}\n\\def\\gpglinetwo{%s}\n", gpg1, gpg2);
      GNUNET_free (gpg2);
      GNUNET_free (gpg1);
    }
    fprintf (f,
             "\\def\\gns{%s/%s}\n",
             gnskey,
             (NULL == gns_nick) ? "" : gns_nick);
    fclose (f);
    GNUNET_asprintf (
      &p,
      "cd %s; cp %s gns-bcd.tex | pdflatex --enable-write18 gns-bcd.tex > /dev/null 2> /dev/null",
      tmp,
      resfile);
    GNUNET_free (deffile);
    ret = system (p);
    if (WIFSIGNALED (ret) || (0 != WEXITSTATUS (ret)))
      GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "system", p);
    GNUNET_asprintf (&deffile, "%s%s%s", tmp, DIR_SEPARATOR_STR, "gns-bcd.pdf");
    fd = open (deffile, O_RDONLY);
    if (-1 == fd)
    {
      GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "open", deffile);
      GNUNET_free (deffile);
      GNUNET_free (p);
      GNUNET_DISK_directory_remove (tmp);
      GNUNET_free (tmp);
      return MHD_NO;
    }
    GNUNET_break (0 == stat (deffile, &st));
    if (NULL ==
        (response = MHD_create_response_from_fd ((size_t) st.st_size, fd)))
    {
      GNUNET_break (0);
      GNUNET_break (0 == close (fd));
      GNUNET_free (deffile);
      GNUNET_free (p);
      GNUNET_DISK_directory_remove (tmp);
      GNUNET_free (tmp);
      return MHD_NO;
    }
    (void) MHD_add_response_header (response,
                                    MHD_HTTP_HEADER_CONTENT_TYPE,
                                    "application/pdf");
    ret = MHD_queue_response (connection, MHD_HTTP_OK, response);
    MHD_destroy_response (response);
    GNUNET_free (deffile);
    GNUNET_free (p);
    GNUNET_DISK_directory_remove (tmp);
    GNUNET_free (tmp);
    return ret;
  }
  return MHD_queue_response (connection,
                             MHD_HTTP_NOT_FOUND,
                             not_found_response);
}


/**
 * Function that queries MHD's select sets and
 * starts the task waiting for them.
 */
static struct GNUNET_SCHEDULER_Task *
prepare_daemon (struct MHD_Daemon *daemon_handle);


/**
 * Call MHD to process pending requests and then go back
 * and schedule the next run.
 */
static void
run_daemon (void *cls)
{
  struct MHD_Daemon *daemon_handle = cls;

  http_task = NULL;
  GNUNET_assert (MHD_YES == MHD_run (daemon_handle));
  http_task = prepare_daemon (daemon_handle);
}


/**
 * Function that queries MHD's select sets and
 * starts the task waiting for them.
 */
static struct GNUNET_SCHEDULER_Task *
prepare_daemon (struct MHD_Daemon *daemon_handle)
{
  struct GNUNET_SCHEDULER_Task *ret;
  fd_set rs;
  fd_set ws;
  fd_set es;
  struct GNUNET_NETWORK_FDSet *wrs;
  struct GNUNET_NETWORK_FDSet *wws;
  int max;
  MHD_UNSIGNED_LONG_LONG timeout;
  int haveto;
  struct GNUNET_TIME_Relative tv;

  FD_ZERO (&rs);
  FD_ZERO (&ws);
  FD_ZERO (&es);
  wrs = GNUNET_NETWORK_fdset_create ();
  wws = GNUNET_NETWORK_fdset_create ();
  max = -1;
  GNUNET_assert (MHD_YES == MHD_get_fdset (daemon_handle, &rs, &ws, &es, &max));
  haveto = MHD_get_timeout (daemon_handle, &timeout);
  if (haveto == MHD_YES)
    tv.rel_value_us = (uint64_t) timeout * 1000LL;
  else
    tv = GNUNET_TIME_UNIT_FOREVER_REL;
  GNUNET_NETWORK_fdset_copy_native (wrs, &rs, max + 1);
  GNUNET_NETWORK_fdset_copy_native (wws, &ws, max + 1);
  ret = GNUNET_SCHEDULER_add_select (GNUNET_SCHEDULER_PRIORITY_HIGH,
                                     tv,
                                     wrs,
                                     wws,
                                     &run_daemon,
                                     daemon_handle);
  GNUNET_NETWORK_fdset_destroy (wrs);
  GNUNET_NETWORK_fdset_destroy (wws);
  return ret;
}


/**
 * Start server offering our hostlist.
 *
 * @return #GNUNET_OK on success
 */
static int
server_start ()
{
  if (0 == port)
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                _ ("Invalid port number %u.  Exiting.\n"),
                port);
    return GNUNET_SYSERR;
  }
  GNUNET_log (GNUNET_ERROR_TYPE_INFO,
              _ ("Businesscard HTTP server starts on %u\n"),
              port);
  daemon_handle = MHD_start_daemon (MHD_USE_DUAL_STACK | MHD_USE_DEBUG,
                                    port,
                                    NULL /* accept_policy_callback */,
                                    NULL,
                                    &access_handler_callback,
                                    NULL,
                                    MHD_OPTION_CONNECTION_LIMIT,
                                    (unsigned int) 512,
                                    MHD_OPTION_PER_IP_CONNECTION_LIMIT,
                                    (unsigned int) 2,
                                    MHD_OPTION_CONNECTION_TIMEOUT,
                                    (unsigned int) 60,
                                    MHD_OPTION_CONNECTION_MEMORY_LIMIT,
                                    (size_t) (16 * 1024),
                                    MHD_OPTION_END);
  if (NULL == daemon_handle)
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                _ ("Could not start businesscard HTTP server on port %u\n"),
                (unsigned int) port);
    return GNUNET_SYSERR;
  }
  http_task = prepare_daemon (daemon_handle);
  return GNUNET_OK;
}


/**
 * Stop HTTP server.
 */
static void
server_stop (void *cls)
{
  (void) cls;
  GNUNET_log (GNUNET_ERROR_TYPE_DEBUG, "HTTP server shutdown\n");
  if (NULL != http_task)
  {
    GNUNET_SCHEDULER_cancel (http_task);
    http_task = NULL;
  }
  if (NULL != daemon_handle)
  {
    MHD_stop_daemon (daemon_handle);
    daemon_handle = NULL;
  }
  if (NULL != main_response)
  {
    MHD_destroy_response (main_response);
    main_response = NULL;
  }
  if (NULL != invalid_gnskey_response)
  {
    MHD_destroy_response (invalid_gnskey_response);
    invalid_gnskey_response = NULL;
  }
  if (NULL != not_found_response)
  {
    MHD_destroy_response (not_found_response);
    not_found_response = NULL;
  }
  if (NULL != resfile)
  {
    GNUNET_free (resfile);
    resfile = NULL;
  }
}


/**
 * Main function that will be run.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param c configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *c)
{
  struct stat st;
  char *dir;
  char *fn;
  int fd;

  (void) cls;
  (void) args;
  (void) cfgfile;
  cfg = c;
  dir = GNUNET_OS_installation_get_path (GNUNET_OS_IPK_DATADIR);
  GNUNET_assert (NULL != dir);
  GNUNET_asprintf (&fn, "%s%s%s", dir, DIR_SEPARATOR_STR, "gns-bcd.html");
  GNUNET_asprintf (&resfile, "%s%s%s", dir, DIR_SEPARATOR_STR, "gns-bcd.tex");
  GNUNET_free (dir);
  fd = open (fn, O_RDONLY);
  if (-1 == fd)
  {
    GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "open", fn);
    GNUNET_free (fn);
    return;
  }
  if (0 != stat (fn, &st))
  {
    GNUNET_log_strerror_file (GNUNET_ERROR_TYPE_ERROR, "open", fn);
    GNUNET_free (fn);
    GNUNET_break (0 == close (fd));
    return;
  }
  GNUNET_free (fn);
  if (NULL ==
      (main_response = MHD_create_response_from_fd ((size_t) st.st_size, fd)))
  {
    GNUNET_break (0);
    GNUNET_break (0 == close (fd));
    return;
  }
  (void) MHD_add_response_header (main_response,
                                  MHD_HTTP_HEADER_CONTENT_TYPE,
                                  "text/html");
  invalid_gnskey_response =
    MHD_create_response_from_buffer (strlen (INVALID_GNSKEY),
                                     INVALID_GNSKEY,
                                     MHD_RESPMEM_PERSISTENT);
  (void) MHD_add_response_header (invalid_gnskey_response,
                                  MHD_HTTP_HEADER_CONTENT_TYPE,
                                  "text/html");
  not_found_response = MHD_create_response_from_buffer (strlen (NOT_FOUND),
                                                        NOT_FOUND,
                                                        MHD_RESPMEM_PERSISTENT);
  (void) MHD_add_response_header (not_found_response,
                                  MHD_HTTP_HEADER_CONTENT_TYPE,
                                  "text/html");
  if (GNUNET_OK != server_start ())
    return;
  GNUNET_SCHEDULER_add_shutdown (&server_stop, NULL);
}


/**
 * The main function for gnunet-gns.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_uint16 ('p',
                                 "port",
                                 "PORT",
                                 gettext_noop (
                                   "Run HTTP serve on port PORT (default is 8888)"),
                                 &port),
    GNUNET_GETOPT_OPTION_END
  };
  int ret;

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;
  GNUNET_log_setup ("gnunet-bcd", "WARNING", NULL);
  ret = (GNUNET_OK ==
         GNUNET_PROGRAM_run (argc,
                             argv,
                             "gnunet-bcd",
                             _ ("GNUnet HTTP server to create business cards"),
                             options,
                             &run,
                             NULL))
        ? 0
        : 1;
  GNUNET_free ((void *) argv);
  return ret;
}


/* end of gnunet-bcd.c */
